### PR TITLE
 Search performance optimization

### PR DESCRIPTION
**Pull request by FishermanZzhang**
We made the following changes:
1. Use raw pointers instead of weak_ptr
2. The shared_ptr is used as memory management, and the raw pointer is used as the real worker
3. Optimized unreasonable function calls
4. parameter references
5. Simple functions are inlined
6. Overwrite BufferedIndexInput::readVInt()

https://github.com/luceneplusplus/LucenePlusPlus/issues/178

@alanw  PTAL

### CODE REVIEW

File: include/lucene++/Array.h [hunk 1]
```diff
@@ -100,7 +100,7 @@ class Array {
     }
 
     TYPE& operator[] (int32_t i) const {
-        BOOST_ASSERT(i >= 0 && i < array->size);
+        // BOOST_ASSERT(i >= 0 && i < array->size);
```
Comment (alanw) [inline include/lucene++/Array.h:103]: Delete

File: include/lucene++/PhraseScorer.h [hunk 2]
```diff
@@ -27,14 +28,18 @@ class PhraseScorer : public Scorer {
...
     ByteArray norms;
     double value;
 
     bool firstTime;
     bool more;
     PhraseQueuePtr pq;
-    PhrasePositionsPtr first;
-    PhrasePositionsPtr last;
+    std::vector<PhrasePositionsPtr> _holds;
+    // PhrasePositionsPtr first;
```
Comment (alanw) [inline include/lucene++/PhraseScorer.h:39]: Delete

File: include/lucene++/SloppyPhraseScorer.h [hunk 3]
```diff
@@ -20,8 +20,10 @@ class SloppyPhraseScorer : public PhraseScorer {
 
 protected:
     int32_t slop;
-    Collection<PhrasePositionsPtr> repeats;
-    Collection<PhrasePositionsPtr> tmpPos; // for flipping repeating pps
+    // Collection<PhrasePositionsPtr> repeats;
```
Comment (alanw) [inline include/lucene++/SloppyPhraseScorer.h:23]: Delete

File: src/core/util/MiscUtils.cpp [hunk 4]
```diff
@@ -124,12 +124,12 @@ bool MiscUtils::equalTypes(const LuceneObjectPtr& first, const LuceneObjectPtr&
     return (typeid(firstRef) == typeid(secondRef));
 }
 
-int64_t MiscUtils::unsignedShift(int64_t num, int64_t shift) {
-    return (shift & 0x3f) == 0 ? num : (((uint64_t)num >> 1) & 0x7fffffffffffffffLL) >> ((shift & 0x3f) - 1);
-}
+// int64_t MiscUtils::unsignedShift(int64_t num, int64_t shift) {
```
Comment (alanw) [inline src/core/util/MiscUtils.cpp:127]: Delete

**Review by alanw — COMMENTED** [review]
This  looks  really great, thank you so much for taking the time to improve  performance!

I've left a few comments around removing comment-out code, but I think it looks good

File: include/lucene++/Array.h (hunk 1 above)
Comment (FishermanZzhang) [reply include/lucene++/Array.h:103]: done

File: include/lucene++/PhraseScorer.h (hunk 2 above)
Comment (FishermanZzhang) [reply include/lucene++/PhraseScorer.h:39]: done

File: include/lucene++/SloppyPhraseScorer.h (hunk 3 above)
Comment (FishermanZzhang) [reply include/lucene++/SloppyPhraseScorer.h:23]: done

File: src/core/util/MiscUtils.cpp (hunk 4 above)
Comment (FishermanZzhang) [reply src/core/util/MiscUtils.cpp:127]: done

**Review by alanw — APPROVED** [review]

**Pull request merged**